### ISSUE TITLE
Add support for javax.tools.JavaCompiler API

### ISSUE THREAD

**Issue by Thihup** [janino-compiler/janino#153]
Both Javac and ECJ implements the java.compiler API, making the change between those two easily, because you can depend on the API instead of the implementation classes.

However, it seems that Janino doesn't provide out-of-the-box implementation for this API, making it a little more complicated to replace the current compiler in some cases.

**Comment by aunkrig** [comment]
Well, the intent of Janino is complimentary: _Provide an API_ that is implemented by several compilers, currently Janino itself and the `java.lang.Compiler`, not the other way round.

(I once attempted to add ECJ, the Eclipse Java compiler, as a third implementation of the Janino API, but gave up at some point,)

But that doesn't mean that what you want cannot be done. Can you please tell me more about your scenario?

**Comment by Thihup** [comment]
@aunkrig Sure. We have an program that translate some source to Java, and then invoke the Java compiler via the java.compiler API. I would like to give the Janino a try to see if using it, our build times would decrease.
As we're using the Java.compiler API, the only two options is the Javac and the ECJ. If Janino implemented too, would be an easy switch.

Maybe it would be possible to use with Maven too.

**Comment by Thihup** [comment]
The main reason would it be because the Java.compiler API is a standard API to compile, making it easy to depend on it and change the implementation if necessary.

**Comment by aunkrig** [comment]
Hey Thiago,

I got your point. Well, the `org.codehaus.janino.Compiler` would probably be faster than JAVAC, but the real boost comes with the `org.codehaus.janino.SimpleCompiler`. The reason being is that the „real“ Java compilers must read and parse lots and lots of .class files, specifically those from the JRE standard library. The trick of `SimpleCompiler` (and its smaller brothers `(Expression|Script|ClassBody)Evaluator` is to instead use the classes that the currently executing JVM has already loaded, which makes the compilation tremendously faster.

However, `java.lang.compiler` is an abstraction for „real“ compilers and cannot be used for the `SimpleCompiler`.

In other words: Yes, it should be possible to wrap Janino in `java.lang.compiler`, and yes, that should be faster, but if you want to compile code over and over, and quickly, then `SimpleCompiler` is the quantum leap.

I recommend that you check the documentation of the `SimpleCompiler` and find out if it suffices. Since you want to compile generated Java code, I assume that the restriction of one single compilation unit is acceptable.

Please let me know what you think.

**Comment by aunkrig** [comment]
Ping.

**Comment by Thihup** [comment]
Thanks!
